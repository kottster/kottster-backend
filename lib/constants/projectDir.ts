// Current project directory
export const PROJECT_DIR = process.cwd();
