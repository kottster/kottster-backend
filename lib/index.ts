export { KottsterApp, KottsterAppOptions } from './core/app';
export { createApp } from './factories/createApp';
export { createAdapter } from './factories/createAdapter';
export { getEnvOrThrow } from './utils/getEnvOrThrow';
export { AdapterService } from './services/adapter.service';
