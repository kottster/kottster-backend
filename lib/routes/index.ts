import { healthcheck } from './healthcheck';
import { executeAction } from './executeAction';
import { executeProcedure } from './executeProcedure';

export default {
  healthcheck,
  executeAction,
  executeProcedure,
}