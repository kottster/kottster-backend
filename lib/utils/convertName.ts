/**
 * Convert a string to human readable format
 * @example 'fooBarBaz' => 'Foo bar baz'
 * @example 'foo_bar_baz' => 'Foo bar baz'
 * @param name The name to convert
 */
export function convertName(name: string): string {
  return name
    .replace(/([a-z])([A-Z])/g, '$1 $2')
    .replace(/_/g, ' ')
    .toLowerCase()
    .split(' ')
    .map((word, i) => i > 0 ? word : word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');
}