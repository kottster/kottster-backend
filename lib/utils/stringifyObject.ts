/**
 * Stringifies an object with single quotes and no double quotes.
 * @param obj The object to stringify
 * @param indent The indentation to use
 */
export function stringifyObject(obj: any, indent: string = ''): string {
  try {
    const replacer = (key: string, value: any) => {
      if (typeof value === 'string') {
        return `'${value.replace(/'/g, "\\'")}'`;
      }
      return value;
    };

    const lines = JSON.stringify(obj, replacer, 2).split('\n');
    const indentedLines = lines.map((line, index) => {
      if (index === 0) {
        return line;
      }
      return indent + line;
    });

    return indentedLines.join('\n').replace(/"([^"]+)":/g, '$1:').replace(/"'/g, "'").replace(/'"/g, "'");
  } catch (error) {
    console.error('Error stringifying object:', error);
    return '';
  }
}