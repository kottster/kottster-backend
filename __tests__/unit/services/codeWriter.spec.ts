import { CodeWriter } from "../../../lib/services/codeWriter.service";
import fs from 'fs';
import mock from 'mock-fs';
import { getSampleProceduresFileContent, getSampleMetricFileProcedure, getSampleAdapterFileContent, getSampleAutoImportFileContent } from "../../utils/samples";
import { createAdapter } from "../../../lib";
import { AdapterType } from "../../../lib/models/adapter.model";
import { PROJECT_DIR } from "../../../lib/constants/projectDir";

const writeFileSyncSpy = jest.spyOn(fs, 'writeFileSync');

describe('services/codeWriter', () => {
  beforeEach(() => {
    mock({
      [`${PROJECT_DIR}/src/dev`]: {},
      [`${PROJECT_DIR}/src/prod`]: {},
      [`${PROJECT_DIR}/src/__generated__/dev/procedures`]: {},
      [`${PROJECT_DIR}/src/__generated__/prod/procedures`]: {},
      [`${PROJECT_DIR}/src/dev/adapter.js`]: 'export {};',
    });
  });
  
  afterEach(() => {
    mock.restore();
  });

  it(`should create file with adapter`, async () => {
    const adapter = createAdapter(AdapterType.postgresql, {
      connectionOptions: {
        connection: {
          host: "localhost",
          port: 5432,
          user: "root",
          password: "password123",
          database: "postgres"
        },
        searchPath: ['public']
      },
    });
    
    const codeWriter = new CodeWriter();
    await codeWriter.writeAdapterToFile(adapter);

    const calledWithArgs = writeFileSyncSpy.mock.calls[0];
    expect(calledWithArgs[1]).toMatchSnapshot();
  });

  it(`should create file with procedures`, async () => {
    const filePath = `${PROJECT_DIR}/src/__generated__/dev/procedures/page_p1_metric_m1.js`;
    
    const codeWriter = new CodeWriter();
    await codeWriter.writeComponentProceduresToFile('p1', 'metric', 'm1', [
      getSampleMetricFileProcedure()
    ]);

    const calledWithArgs = writeFileSyncSpy.mock.calls[2];
    expect(calledWithArgs[1]).toMatchSnapshot();
  });

  it(`should generate autoimports`, async () => {
    const codeWriter = new CodeWriter();
    await codeWriter.generateAutoImports();

    const calledWithArgs = writeFileSyncSpy.mock.calls[3];
    expect(calledWithArgs[1]).toMatchSnapshot();
  });
});