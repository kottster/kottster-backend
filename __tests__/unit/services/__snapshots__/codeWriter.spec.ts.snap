// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`services/codeWriter should create file with adapter 1`] = `
"import { createAdapter } from '@kottster/backend';

const adapters = [
  createAdapter(
    'postgresql', 
    {
      connectionOptions: {
        connection: {
          host: 'localhost',
          port: 5432,
          user: 'root',
          password: 'password123',
          database: 'postgres'
        },
        searchPath: [
          'public'
        ]
      }
    }
  )
];

export default adapters;

"
`;

exports[`services/codeWriter should create file with procedures 1`] = `
"// This file was generated by the Kottster App Builder
// PLEASE AVOID EDITING THIS FILE MANUALLY

import { app } from '../../../app';

const STAGE = 'dev';
const PAGE_ID = 'p1';
const COMPONENT_TYPE = 'metric';
const COMPONENT_ID = 'm1';

app.registerProceduresForComponent(STAGE, PAGE_ID, COMPONENT_TYPE, COMPONENT_ID, {
  'getUserCount': async function () {return { userCount: 123 };},
});

"
`;

exports[`services/codeWriter should generate autoimports 1`] = `
"// This file was generated by the Kottster App Builder
// PLEASE AVOID EDITING THIS FILE MANUALLY

import './dev/procedures/page_p1_metric_m1.js';
"
`;
